keps/sig-windows/20190714-windows-csi-support.md

```diff
@@ -0,0 +1,401 @@
+
+The API versioning scheme described above, as of now, will maintain backward compatibility indefinitely. This requires the scope of csi-proxy.exe to be limited to a very scoped down fundamental set of operations. Maintainers therefore will need to be very cautious when accepting suggestions for new APIs and enhancements. This may slow progress at times.
+
+There may ultimately be certain operations that csi-proxy.exe cannot support in an elegant fashion and requires the plugin author targeting Windows nodes to seek one of the alternatives described below. This KEP and maintainers should ensure such instances are as limited as possible.
+
+## Alternatives
+
+There are multiple alternatives to deploying CSI Node Plugins as containers along with csi-proxy.exe for privileged operations however each has it's own set of disadvantages.
+
+### Deploying CSI Node Plugins as binaries and deployed as processes running on the host:
+
+With this approach, lifecycle of multiple stand-alone binaries corresponding to different CSI node plugins will need to be managed. The standard CSI Node Driver Registrar which is distributed as a container will also need to be repackaged as binaries and distributed (as mixing side car containers with standalone binaries is not possible). Managing several binaries outside of Kubernetes may not scale well as diverse storage systems, each with their own CSI plugin, is integrated with the cluster.
+
+### Package CSI Node Plugins as containers and deployed as processes running on the host:
+
+With this approach, the container run time is enhanced to be able to launch binaries directly on the host after pulling down a container image and extracting the binary from the image. While usual Kubernetes constructs may be used to launch pods with a special RuntimeClass that can handle launching of the binaries as processes on hosts, Kubernetes won't be able to fully monitor and control the whole life-cycle of the binaries post launch. Collection of logs from the plugins also become problematic and will require tooling out-of-band from Kubernetes.
+
+### Support for Privileged Operations and Bi-directional mount propagation in Windows containers:
+
+At some point in the future, it is expected that a Windows LTS release will implement support for execution of privileged operations and bi-directional mount propagation from inside containers. At that point, the requirement of a proxy to handle privileged operations on behalf of containers will disappear.
+
+## Infrastructure Needed
+
+The code for csi-proxy as well as the GRPC API will be maintained in a dedicated repo: github.com/kubernetes-csi/windows-csi-proxy 
```

Comment:
Will there be a csi proxy for linux as well?

Reply:
At a high level, a CSI proxy for Linux in the context of operations enumerated in the API above, is unnecessary as the operations can be executed from privileged containers directly in Linux. However, as @msau42 pointed out above there are some special scenarios in Linux where a host footprint may be necessary for Linux plugins as well. iSCSI is an interesting corner case today for Linux where the `iscsiadm` binary is typically mounted into the plugin container from the host while interactions with the iSCSI kernel modules is facilitated through `iscsid` running on the host. 

I will call out that the proxy called out in the KEP focusses on Windows today and may be expanded for Linux scenarios in the future. Does that sound good @msau42 @yujuhong?